### PR TITLE
moved Diagnostic usage to DiagnosticData in workspace layer

### PR DESCRIPTION
**Pull request by heejaechang**
also moved DiagnosticData down to Workspace layer

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani can you take a look? ignore project.json stuff, I will take those out.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs [hunk 1]
```diff
@@ -368,19 +368,15 @@ private static IEnumerable<SuggestedActionSet> PrioritizeFixGroups(IDictionary<D
                     var fixes = map[diag];
 
                     var priority = fixes.All(s => s is SuppressionSuggestedAction) ? SuggestedActionSetPriority.None : SuggestedActionSetPriority.Medium;
-                    var applicableToSpan = GetApplicableToSpan(diag);
 
-                    sets.Add(new SuggestedActionSet(fixes, priority, applicableToSpan));
+                    // diagnostic from things like build shouldn't reach here since we don't support LB for those diagnostics
+                    Contract.Requires(diag.HasTextSpan);
```
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs]: This might be true now, but we now support suppressing no-location diagnostics from error list, and when we add LB support for code fixes from error list it might no longer be true - but that presumable would need lot more rewrite of the code in this file.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuppressionSuggestedAction.cs [hunk 2]
```diff
@@ -119,7 +119,7 @@ public string GetDiagnosticID()
             }
 
             // if it is from third party, we use hashcode
-            return diagnostic.GetHashCode().ToString(CultureInfo.InvariantCulture);
+            return diagnostic.Id.GetHashCode().ToString(CultureInfo.InvariantCulture);
```
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuppressionSuggestedAction.cs]: Why did we change this?

Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuppressionSuggestedAction.cs]: diagnostic.GetHashCode() contains location which can change as you change unrelated code. which make telemetry data pretty much useless.

File: src/Features/Core/Portable/CodeFixes/CodeFixService.cs [hunk 3]
```diff
@@ -277,8 +274,10 @@ internal partial class CodeFixService : ICodeFixService
                 FixAllCodeActionContext fixAllContext = null;
                 if (fixAllProviderInfo != null)
                 {
-                    var codeFixProvider = (fixer as CodeFixProvider) ?? new WrapperCodeFixProvider((ISuppressionFixProvider)fixer, diagnostics);
-                    fixAllContext = FixAllCodeActionContext.Create(document, fixAllProviderInfo, codeFixProvider, diagnostics, this.GetDocumentDiagnosticsAsync, this.GetProjectDiagnosticsAsync, cancellationToken);
+                    var codeFixProvider = (fixer as CodeFixProvider) ?? new WrapperCodeFixProvider((ISuppressionFixProvider)fixer, diagnostics.Select(d => d.Id).Distinct().ToImmutableArray());
```
Comment (mavasani) [inline src/Features/Core/Portable/CodeFixes/CodeFixService.cs]: Why did we move `.Select(d => d.Id).Distinct().` invocation out of the WrapperCodeFixProvider ctor? All the callers now need to invoke this right?

File: src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs [hunk 4]
```diff
@@ -12,10 +12,10 @@ internal sealed class WrapperCodeFixProvider : CodeFixProvider
         private readonly ImmutableArray<string> _originalDiagnosticIds;
         private readonly ISuppressionFixProvider _suppressionFixProvider;
 
-        public WrapperCodeFixProvider(ISuppressionFixProvider suppressionFixProvider, ImmutableArray<Diagnostic> originalDiagnostics)
+        public WrapperCodeFixProvider(ISuppressionFixProvider suppressionFixProvider, ImmutableArray<string> diagnosticIds)
```
Comment (mavasani) [inline src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs]: Can we instead just change this to take ImmutableArray<DiagnosticData> as the parameter and retain the `.Select(d => d.Id).Distinct()` invocation here?

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal static class Extensions
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Diagnostics/Extensions.cs]: I am assuming this is just moving same APIs down to Workspace layer with no functionality changes, is that correct?

**Comment by mavasani** [comment]
:+1:

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs (hunk 1 above)
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs]: yep. probably once we get there. how suppression one works now?

File: src/Features/Core/Portable/CodeFixes/CodeFixService.cs (hunk 3 above)
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/CodeFixService.cs]: to break dependency to Diagnostics from WrapperCodeFixProvider. only reason it wanted to have Diagnostic was to get ID list. because of it all caller must have Diagnostics making caller that only has DiagnosticData to be required to convert them to Diagnostic first and then get ID and then throw Diagnostics away.

File: src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs (hunk 4 above)
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs]: that means, again caller who has Diagnostics only need to convert them to DiagnosticData and then get ID and then through them away. the API only need ID so it should accept just id. we can create helper to get those if you think Select(..) is duplication.

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Diagnostics/Extensions.cs]: yep. I just moved this out of DiagnosticData.

File: src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs (hunk 4 above)
Comment (mavasani) [inline src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs]: okay makes sense, probably you can leave the .Distinct() call inside the ctor as the caller really shouldn't need to worry about it?

Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/WrapperCodeFixProvider.cs]: Will do that.


-          heejae

**Comment by heejaechang** [comment]
@dotnet-bot retest eta please

**Comment by heejaechang** [comment]
@dotnet-bot retest win please

**Comment by heejaechang** [comment]
@dotnet-bot retest unit please

**Comment by heejaechang** [comment]
@dotnet-boat retest unit please

**Comment by heejaechang** [comment]
@dotnet-boat retest unit please

**Comment by heejaechang** [comment]
@dotnet-bot retest linux please

**Comment by heejaechang** [comment]
@dotnet-bot retest this please

**Comment by heejaechang** [comment]
this fixed https://github.com/dotnet/roslyn/issues/3111
